lib/Conversion/TorchOnnxToTorch/OnnxLstmExpander.cpp

```diff
@@ -0,0 +1,479 @@
+    Value cstZero = rewriter.create<Torch::ConstantIntOp>(
+        loc, intType, rewriter.getIntegerAttr(rewriter.getIntegerType(64), 0));
+    Value cstDirection = rewriter.create<Torch::ConstantIntOp>(
+        loc, intType,
+        rewriter.getIntegerAttr(rewriter.getIntegerType(64), direction));
+
+    return rewriter.create<Torch::AtenSelectIntOp>(loc, outputType, input,
+                                                   cstZero, cstDirection);
+  };
+
+  Value W_forward = getDirection(0, W);
+  Value R_forward = getDirection(0, R);
+  Value B_forward = getDirection(0, B);
+
+  Torch::ValueTensorType HType = rewriter.getType<Torch::ValueTensorType>(
+      llvm::SmallVector<int64_t>{num_directions, batch_size, hidden_size},
+      XType.getDtype());
+
+  auto intType = rewriter.getType<Torch::IntType>();
+
+  // construct a list containing the shape of initial_h and initial_c
+  // this is used to check if initial_h and initial_c are provided
+  Value cstNumDirections = rewriter.create<Torch::ConstantIntOp>(
+      loc, intType,
+      rewriter.getIntegerAttr(rewriter.getIntegerType(64), num_directions));
+  Value cstBatchSize = rewriter.create<Torch::ConstantIntOp>(
+      loc, intType,
+      rewriter.getIntegerAttr(rewriter.getIntegerType(64), batch_size));
+  Value cstHiddenSize = rewriter.create<Torch::ConstantIntOp>(
+      loc, intType,
+      rewriter.getIntegerAttr(rewriter.getIntegerType(64), hidden_size));
+  Value cstNone = rewriter.create<Torch::ConstantNoneOp>(loc);
+  Value cstZero = rewriter.create<Torch::ConstantIntOp>(
+      loc, intType, rewriter.getIntegerAttr(rewriter.getIntegerType(64), 0));
+  Value cstOne = rewriter.create<Torch::ConstantIntOp>(
+      loc, intType, rewriter.getIntegerAttr(rewriter.getIntegerType(64), 1));
+
+  Value HShape = rewriter.create<Torch::PrimListConstructOp>(
+      loc, rewriter.getType<Torch::ListType>(intType),
+      ValueRange({cstNumDirections, cstBatchSize, cstHiddenSize}));
+
+  Value cstDtype =
+      Torch::getDtypeIntValueForType(rewriter, loc, XType.getDtype());
+
+  // initialize hidden and cell states
+  if (binder.tensorOperandAtIndex(initial_h, 5)) {
+    initial_h = rewriter.create<Torch::AtenZerosOp>(
+        loc, HType, HShape, cstDtype, cstNone, cstNone, cstNone);
+  }
+  if (binder.tensorOperandAtIndex(initial_c, 6)) {
+    initial_c = rewriter.create<Torch::AtenZerosOp>(
+        loc, HType, HShape, cstDtype, cstNone, cstNone, cstNone);
+  }
+
+  Value initial_h_forward = getDirection(0, initial_h);
+  Value initial_c_forward = getDirection(0, initial_c);
+
+  // ### everything hereon is only one direction. they won't have the direction
+  // dimension. todo: support bidirectional and reverse LSTM,
+  // possibly by doing each direction separately and combining
+  Value HSizeX4 = rewriter.create<Torch::ConstantIntOp>(
+      loc, intType,
+      rewriter.getIntegerAttr(rewriter.getIntegerType(64), 4 * hidden_size));
```

Comment:
Details on why 4 and 8 are important. `constexpr` can make this a lot cleaner.

Reply:
Could you give me an example of how to use constexpr to make this cleaner? Like a constexpr function?